core/chainio/avs_writer.go

```diff
@@ -87,21 +100,61 @@ func (w *AvsWriter) SendAggregatedResponse(batchIdentifierHash [32]byte, batchMe
 		return nil, err
 	}
 
-	// Send the transaction
+	// Set the nonce, as we might have to replace the transaction with a higher fee
+	txNonce := new(big.Int).SetUint64(tx.Nonce())
 	txOpts.NoSend = false
-	txOpts.GasLimit = tx.Gas() * 110 / 100 // Add 10% to the gas limit
-	tx, err = w.AvsContractBindings.ServiceManager.RespondToTaskV2(&txOpts, batchMerkleRoot, senderAddress, nonSignerStakesAndSignature)
-	if err != nil {
-		// Retry with fallback
-		tx, err = w.AvsContractBindings.ServiceManagerFallback.RespondToTaskV2(&txOpts, batchMerkleRoot, senderAddress, nonSignerStakesAndSignature)
+	txOpts.Nonce = txNonce
+
+	i := 0
+	sendTransaction := func() (*types.Receipt, error) {
+		if i > 0 {
+			onRetry()
+		}
+		i++
+
+		// get gas price
+		gasPrice, err := w.Client.SuggestGasPrice(context.Background())
+		if err != nil {
+			// Retry with fallback
+			gasPrice, err = w.ClientFallback.SuggestGasPrice(context.Background())
+			if err != nil {
+				return nil, fmt.Errorf("transaction simulation failed: %v", err)
+			}
+		}
+		txOpts.GasPrice = utils.CalculateGasPriceBumpBasedOnRetry(gasPrice, GasBumpPercentage, i)
+
+		w.logger.Infof("Sending ResponseToTask transaction with a gas price of %v", txOpts.GasPrice)
+		err = w.checkRespondToTaskFeeLimit(tx, txOpts, batchIdentifierHash, senderAddress)
+
+		if err != nil {
+			return nil, connection.PermanentError{Inner: err}
+		}
+
+		tx, err = w.AvsContractBindings.ServiceManager.RespondToTaskV2(&txOpts, batchMerkleRoot, senderAddress, nonSignerStakesAndSignature)
+		if err != nil {
+			// Retry with fallback
+			tx, err = w.AvsContractBindings.ServiceManagerFallback.RespondToTaskV2(&txOpts, batchMerkleRoot, senderAddress, nonSignerStakesAndSignature)
+			if err != nil {
+				return nil, connection.PermanentError{Inner: err}
```

Comment:
RPC errors are marked as permanent to not fall into an infinite retry. Waiting for the receipt and not getting is always a recoverable error as we are bumping the fee, at some point our tx should be accepted. That is why this function is wrapped in another retry that has a max which would account for rpc errors.

Reply:
See [here](https://github.com/yetanotherco/aligned_layer/blob/f0e58cf1545c566d0b4e61af1167561656bdf8b5/aggregator/internal/pkg/aggregator.go#L272-L286).

Reply:
So as we talked, we should have the "normal error" (a simple return err) , and errors that don't bump the fee; instead of permanent errors which give place to confusion (more so when the outside loop is removed)